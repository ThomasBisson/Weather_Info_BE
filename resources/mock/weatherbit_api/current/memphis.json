[
	{
		"app_temp":10.1,
		"aqi":42,
		"city_name":"Memphis",
		"clouds":40,
		"country_code":"US",
		"datetime":"2023-12-03:16",
		"dewpt":9.2,
		"dhi":98.88,
		"dni":818.28,
		"elev_angle":31.51,
		"ghi":519.89,
		"gust":null,
		"h_angle":-18,
		"lat":35.14953,
		"lon":-90.04898,
		"ob_time":"2023-12-03 16:39",
		"pod":"d",
		"precip":0,
		"pres":1006.6,
		"rh":94,
		"slp":1014.9,
		"snow":0,
		"solar_rad":499.1,
		"sources":[
			"KM01",
			"radar",
			"satellite"
		],
		"state_code":"TN",
		"station":"KM01",
		"sunrise":"12:52",
		"sunset":"22:48",
		"temp":10.1,
		"timezone":"America/Chicago",
		"ts":1701621593,
		"uv":2.1621995,
		"vis":13,
		"weather":{
			"icon":"c02d",
			"description":"Few clouds",
			"code":801
		},
		"wind_cdir":"W",
		"wind_cdir_full":"west",
		"wind_dir":270,
		"wind_spd":2.57
	}
  ]