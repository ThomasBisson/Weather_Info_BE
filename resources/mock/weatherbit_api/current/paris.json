[
	{
    	"app_temp": 0.1,
    	"aqi": 82,
    	"city_name": "Paris",
    	"clouds": 53,
    	"country_code": "FR",
    	"datetime": "2023-12-03:15",
    	"dewpt": -1.1,
    	"dhi": 41.02,
    	"dni": 359.18,
    	"elev_angle": 6.24,
    	"ghi": 70.62,
    	"gust": 5.14,
    	"h_angle": 90,
    	"lat": 48.85341,
    	"lon": 2.3488,
    	"ob_time": "2023-12-03 15:22",
    	"pod": "d",
    	"precip": 0,
    	"pres": 1009.5,
    	"rh": 70,
    	"slp": 1017,
    	"snow": 0,
    	"solar_rad": 64.1,
    	"sources": [ "analysis", "C1292", "radar", "satellite" ],
    	"state_code": "11",
    	"station": "C1292",
    	"sunrise": "07:25",
    	"sunset": "15:56",
    	"temp": 3.3,
    	"timezone": "Europe/Paris",
    	"ts": 1701616933,
    	"uv": 0.8046274,
    	"vis": 16,
    	"weather": {
			"code": 802,
			"description": "Scattered clouds",
			"icon": "c02d"
		},
    	"wind_cdir": "S",
    	"wind_cdir_full": "south",
    	"wind_dir": 175,
    	"wind_spd": 3.6
	}
]