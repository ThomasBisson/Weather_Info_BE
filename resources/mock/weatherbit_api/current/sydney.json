[
	{
		"app_temp":19.5,
		"aqi":13,
		"city_name":"Sydney",
		"clouds":63,
		"country_code":"AU",
		"datetime":"2023-12-03:16",
		"dewpt":15.9,
		"dhi":0,
		"dni":0,
		"elev_angle":-17.28,
		"ghi":0,
		"gust":3.6,
		"h_angle":-90,
		"lat":-33.86785,
		"lon":151.20732,
		"ob_time":"2023-12-03 16:53",
		"pod":"n",
		"precip":0,
		"pres":1010,
		"rh":78,
		"slp":1010.11847,
		"snow":0,
		"solar_rad":0,
		"sources":[
			"analysis",
			"D5330",
			"radar",
			"satellite"
		],
		"state_code":"02",
		"station":"D5330",
		"sunrise":"18:36",
		"sunset":"08:53",
		"temp":19.4,
		"timezone":"Australia/Sydney",
		"ts":1701622389,
		"uv":0,
		"vis":16,
		"weather":{
			"icon":"c02n",
			"description":"Scattered clouds",
			"code":802
		},
		"wind_cdir":"SSE",
		"wind_cdir_full":"south-southeast",
		"wind_dir":148,
		"wind_spd":3.6
	}
  ]