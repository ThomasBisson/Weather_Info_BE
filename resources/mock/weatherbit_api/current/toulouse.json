[
	{
		"app_temp": 9,
		"aqi": 91,
		"city_name": "Toulouse",
		"clouds": 40,
		"country_code": "FR",
		"datetime": "2023-12-03:15",
		"dewpt": -4.3,
		"dhi": 55.59,
		"dni": 502.24,
		"elev_angle": 10.37,
		"ghi": 137.44,
		"gust": null,
		"h_angle": 90,
		"lat": 43.60426,
		"lon": 1.44367,
		"ob_time": "2023-12-03 15:13",
		"pod": "d",
		"precip": 0,
		"pres": 999.3,
		"rh": 39,
		"slp": 1018,
		"snow": 0,
		"solar_rad": 132.6,
		"sources": [ "LFBO", "radar", "satellite" ],
		"state_code": "76",
		"station": "LFBO",
		"sunrise": "07:10",
		"sunset": "16:18",
		"temp": 9,
		"timezone": "Europe/Paris",
		"ts": 1701616416,
		"uv": 0.9641106,
		"vis": 16,
		"weather": {
		"description": "Few clouds",
		"code": "801",
		"icon": "c02d"
		},
		"wind_cdir": "E",
		"wind_cdir_full": "east",
		"wind_dir": 90,
		"wind_spd": 4.6
	}
  ]