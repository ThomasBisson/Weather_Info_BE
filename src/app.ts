import express, { Application, Request, Response } from 'express';
import cors from 'cors';
import helmet from 'helmet';

import swagger from './swagger';

import errorHandling from './middlewares/errorHandling.middleware';
import morganMiddleware from './middlewares/morgan.middleware';

import weatherRouter from './routes/weather.router';

const app: Application = express();

app.use(morganMiddleware);
app.use(helmet());
app.use(cors());

// NOTE : Use express.json() or body-parser if you need to handle data in the request body
//        For now it's not needed so I commented as it slows the API for no good reason
// App.use(express.json());

// Swagger
if (process.env.NODE_ENV === 'development') {
	swagger(app);
}

// Routes
app.get('/', (req: Request, res: Response) => {
	res.json({ response: 'Welcome to the weather info API !' });
});

app.use('/weather', weatherRouter);

// 404 route
app.get('*', (req, res) => {
	res.status(404).json({
		error: `Cannot find ${req.method} ${req.path}`,
	});
});

// Error handler
app.use(errorHandling);

export default app;
