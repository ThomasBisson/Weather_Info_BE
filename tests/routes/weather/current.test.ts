import { describe, expect, jest, test } from '@jest/globals';
import request from 'supertest';

import app from './../../../src/app';

import * as currentController from '../../../src/controllers/weather/current.controller';

import { WeatherbitCurrentData } from '../../../src/types/weatherbit.type';

import { readMockJsonFile } from '../../../src/utils/file.util';

const spyGetWeatherbitData = jest.spyOn(currentController, 'getWeatherbitData');

describe('Integration /weather/current', () => {
	test('return status 200 and result is OK for paris', async () => {
		const city = 'paris';
		const data = readMockJsonFile(city, 'current') as WeatherbitCurrentData[];
		spyGetWeatherbitData.mockReturnValue(Promise.resolve(data));

		const res = await request(app).get(`/weather/current?city=${city}`);

		expect(res.statusCode).toBe(200);

		expect(res.body).toMatchSnapshot();
	});

	test('return status 200 and result is OK for toulouse', async () => {
		const city = 'toulouse';
		const data = readMockJsonFile(city, 'current') as WeatherbitCurrentData[];
		spyGetWeatherbitData.mockReturnValue(Promise.resolve(data));

		const res = await request(app).get(`/weather/current?city=${city}`);

		expect(res.statusCode).toBe(200);

		expect(res.body).toMatchSnapshot();
	});

	test('return status 200 and result is OK for memphis', async () => {
		const city = 'memphis';
		const data = readMockJsonFile(city, 'current') as WeatherbitCurrentData[];
		spyGetWeatherbitData.mockReturnValue(Promise.resolve(data));

		const res = await request(app).get(`/weather/current?city=${city}`);

		expect(res.statusCode).toBe(200);

		expect(res.body).toMatchSnapshot();
	});

	test('return status 200 and result is OK for sydney', async () => {
		const city = 'sydney';
		const data = readMockJsonFile(city, 'current') as WeatherbitCurrentData[];
		spyGetWeatherbitData.mockReturnValue(Promise.resolve(data));

		const res = await request(app).get(`/weather/current?city=${city}`);

		expect(res.statusCode).toBe(200);

		expect(res.body).toMatchSnapshot();
	});
});
