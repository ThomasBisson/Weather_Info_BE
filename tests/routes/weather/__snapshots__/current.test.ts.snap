// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration /weather/current return status 200 and result is OK for memphis 1`] = `
{
  "US_TN": {
    "description": "Few clouds",
    "relativeHumidity": 94,
    "temperature": 10.1,
    "windSpeed": 9.251999999999999,
  },
}
`;

exports[`Integration /weather/current return status 200 and result is OK for paris 1`] = `
{
  "FR_11": {
    "description": "Scattered clouds",
    "relativeHumidity": 70,
    "temperature": 3.3,
    "windSpeed": 12.96,
  },
}
`;

exports[`Integration /weather/current return status 200 and result is OK for sydney 1`] = `
{
  "AU_02": {
    "description": "Scattered clouds",
    "relativeHumidity": 78,
    "temperature": 19.4,
    "windSpeed": 12.96,
  },
}
`;

exports[`Integration /weather/current return status 200 and result is OK for toulouse 1`] = `
{
  "FR_76": {
    "description": "Few clouds",
    "relativeHumidity": 39,
    "temperature": 9,
    "windSpeed": 16.56,
  },
}
`;
